File: 0743-networkDelayTime.js
// https://leetcode.com/problems/network-delay-time/

function networkDelayTime(times, N, K) {
    const graph = createAdjacencyMap(times, N);
    const dists = { [K]: 0 };
    const mpq = new MinPriorityQueue(Object.keys(graph));
    mpq.set(K, 0);
    
    while (!mpq.isEmpty()) {
        const entry = mpq.popEntry();
        const name = entry.key;
        const dist = entry.val;
        
        const children = graph[name];
        for (const child of children) {
            if (dist + child.dist < mpq.get(child.name)) {
                mpq.set(child.name, dist + child.dist);
                dists[child.name] = dist + child.dist;
            }
        }
    }
        
    if (Object.keys(dists).length < N) {
        return -1;
    }
    
    let maxDist = 0;
    
    for (const name in dists) {
        const dist = dists[name];
        maxDist = Math.max(maxDist, dist);
    }
    
    return maxDist;
}

function createAdjacencyMap(times, N) {
    const graph = {};
    
    for (let i = 1; i <= N; i++) {
        graph[i] = [];
    }
    
    for (const time of times) {
        const [u, v, w] = time;
        graph[u].push({ name: v, dist: w });
    }
    
    return graph;
}

class MinPriorityQueue {
    constructor(keys) {
        this.map = {};
        for (const key of keys) {
            this.map[key] = Infinity;
        }
    }
    
    get(key) {
        return this.map[key];
    }
    
    set(key, val) {
        this.map[key] = val;
    }
    
    isEmpty() {
        return Object.keys(this.map).length === 0;
    }
    
    popEntry() {
        let minKey = null;
        let minVal = Infinity;
        
        for (const key in this.map) {
            const val = this.map[key];
            
            if (val <= minVal) {
                minKey = key;
                minVal = val;
            }
        }
        
        delete this.map[minKey];
        return { key: minKey, val: minVal };
    }
}

File: java/0743-networkDelayTime.java
// https://leetcode.com/problems/network-delay-time/

class Solution {
    public int networkDelayTime(int[][] times, int N, int K) {
        // init adjacency list, priority queue w/ id lookup, and best total distances
        
        Map<Integer, List<Edge>> graph = new HashMap<>();
        PriorityQueue<Path> queue = new PriorityQueue<>();
        Map<Integer, Path> index = new HashMap<>();
        Map<Integer, Integer> distances = new HashMap<>();
        
        for (int id = 1; id <= N; id++) {
            Path path = new Path(id, K);
            
            graph.put(id, new ArrayList<>());
            queue.add(path);
            index.put(id, path);
        }
        
        for (int[] time : times) {
            int id = time[0];
            graph.get(id).add(new Edge(time));
        }
        
        // Disjkstra's algorithm
        
        while (!queue.isEmpty()) {            
            Path curr = queue.poll();
            index.remove(curr.id);
                        
            distances.put(curr.id, curr.dist);
                        
            for (Edge next : graph.get(curr.id)) {
                Path best = index.get(next.id);
                
                if (best != null && curr.dist + next.dist < best.dist) {
                    queue.remove(best);
                    best.dist = curr.dist + next.dist;
                    queue.add(best); // necessary to reprioritize queue
                }
            }
        }
        
        if (distances.values().stream().anyMatch(d -> d == Integer.MAX_VALUE)) {
            return -1;
        } else {
            return Collections.max(distances.values());
        }
    }
    
    private class Path implements Comparable<Path> {
        public int id;
        public int dist;
        
        public Path(int id, int K) {
            this.id = id;
            this.dist = (id == K)
                ? 0
                : Integer.MAX_VALUE;
        }
        
        @Override
        public int compareTo(Path that) {
            return Integer.compare(this.dist, that.dist);
        }
    }
    
    private class Edge {
        public int id;
        public int dist;
        
        public Edge(int[] time) {
            this.id = time[1];
            this.dist = time[2];
        }
    }
}